### ISSUE TITLE
Cannot parse type signature containing constraint with double tilde (~~)

### ISSUE THREAD

**Issue by alexpeits** [ndmitchell/hlint#831]
hlint fails with a parse error on this piece of code (whereas ghc/ghci are happy):

``` haskell
{-# LANGUAGE GADTs          #-}
{-# LANGUAGE KindSignatures #-}
{-# LANGUAGE PolyKinds      #-}
{-# LANGUAGE RankNTypes     #-}
{-# LANGUAGE TypeOperators  #-}

import Data.Kind          (Type)
import Data.Type.Equality

hgcastWith
  :: forall (a :: k) (b :: k') (r :: Type).
     (a :~~: b)
  -> (a ~~ b => r)
  -> r
hgcastWith HRefl x = x
```

```
hlint-reprod.hs:13:14: Error: Parse error
Found:
      :: forall (a :: k) (b :: k') (r :: Type).
         (a :~~: b)
  >   -> (a ~~ b => r)
      -> r
    hgcastWith HRefl x = x
  

1 hint
```

If I change the offending line to `(a ~ b =>  r)` or `(a ~~ b -> r)` hlint checks pass. The second one doesn't really make sense but it makes me think that it's the combination of `~~` and `=>` that makes it fail.

I've tried this with multiple hlint versions, currently with `v2.2.4` and also latest `master`. I can also parse it using `ghc-lib-parser` so maybe it's because of `haskell-src-exts`?

**Comment by ndmitchell** [comment]
Thanks. We're in the middle of replacing the parser, so will wait til that is done and see the impact it had.

**Comment by ndmitchell** [comment]
Confirmed works with HEAD that doesn't use HSE for parsing, so fixed in the next release.
